03week/ticTacToe.js

```diff
@@ -23,32 +46,91 @@ function printBoard() {
   console.log('2 ' + board[2].join(' | '));
 }
 
-function horizontalWin() {
-  // Your code here
+const isValidInput = (row, column) => {
+  const validInputs = [0, 1, 2];
+  const isValid = validInputs.indexOf(row) > -1 || validInputs.indexOf(column) > -1
+  return isValid
+}
+
+const isOpenForPlay = (row, column) => {
+  return board[row][column] === ' '
+}
+
+const checksArrayForPlayerWin = (arr) => {
+  return arr.every((item) => item === playerTurn)
+}
+
+const checkMatrixForWin = (board) => {
+  let results = board.filter(checksArrayForPlayerWin);
+  return results.length > 0
+}
+
+const horizontalWin = () => {
+  return checkMatrixForWin(board);
 }
 
-function verticalWin() {
-  // Your code here
+const rotateMatrix = (matrix) => {
+   let rotatedMatrix = []
+   for (var x = 0; x < matrix.length; x++) {
+       rotatedMatrix[x] = []
+       for (var y = 0; y < matrix.length; y++) {
+           rotatedMatrix[x][y] = matrix[y][x]
+       }
+   }
+   return rotatedMatrix
 }
 
-function diagonalWin() {
-  // Your code here
+const verticalWin = () => {
+  return checkMatrixForWin(rotateMatrix(board))
+}
+
+const diagonalWin = () => {
+  const diagonalDown = [board[0][0], board[1][1], board[2][2]];
+  const diagonalUp = [board[2][0], board[1][1], board[0][2]];
+  return checkMatrixForWin([diagonalDown, diagonalUp]);
+}
+
+const switchPlayer = () => {
+  if(playerTurn === 'X'){
+    playerTurn = 'O';
+  } else {
+    playerTurn = 'X'
+  }
 }
 
 function checkForWin() {
-  // Your code here
+  const horizontal = horizontalWin();
+  const vertical = verticalWin();
+  const diagonal = diagonalWin();
+  return horizontal || vertical || diagonal
 }
 
 function ticTacToe(row, column) {
-  // Your code here
+  if(!isValidInput(row, column)){
+    return console.log('invalid entry, try again!')
+  }
+  if (!isOpenForPlay(row, column)) {
+    return console.log('Invalid move, try again!')
+  }
+  board[row][column] = playerTurn;
+  incrementCounter();
+  if (checkForWin()) {
+    console.log(`Player ${playerTurn} won!`)
+    board = resetBoard();
+  } 
+  if(catsScratch()){
+    console.log("It's a tie! Try again!")
+    board = resetBoard();
+  }
+  switchPlayer();
 }
 
 function getPrompt() {
   printBoard();
   console.log("It's Player " + playerTurn + "'s turn.");
   rl.question('row: ', (row) => {
     rl.question('column: ', (column) => {
-      ticTacToe(row, column);
+      ticTacToe(parseInt(row), parseInt(column));
```

Comment:
This is a great practice on passing argument to a function call, However understand that the method parseInt can only take in a number if a different character is received it would return NaN.